### PR TITLE
Automatically Assign Goalie in STP and Fix STP Assignment Bug

### LINKED ISSUE #1859: Automatically Assign GoalieTactic to the Goalie in STP
### Description of the task
The GoalieTactic must always be assigned to the goalie, so it's not necessary for every play to manually include it. Plays should be responsible for calculating Tactics for all the friendly robots other than the goalie and STP can automatically assign the GoalieTactic to the goalie. This does require some thought about when we might want to pull the goalie, so a mini-design doc/discussion is warranted. The benefit of this restructure is that we reduce some of the boilerplate in creating plays.

### Acceptance criteria

- [ ] Assign goalie tactic to goalie in stp
- [ ] design for pulling the goalie

### LINKED ISSUE #2020: Robots don't stop when there is no tactic to assign
### Description of the task
When we have more robots than  tactics, the robots that do not have a tactic assigned should stop moving.

### Acceptance criteria

- [ ] Figure out why the robots aren't stopping
- [ ] Stop the robots
- [ ] Remove placeholder tactics in CornerKickPlay

### Blocked By
N/A

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* automatically assign goalie in STP so Plays don't have to
* Fix STP assignment bug from tiered tactic assignment
* execute stop tactics and goalie tactic created by STP

### Testing Done
* added regression test for STP assignment bug
* updated all affected tests

### Resolved Issues
resolves #2020
resolves #1859 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/stp.cpp [hunk 1]
```diff
@@ -87,7 +94,23 @@ std::vector<std::unique_ptr<Intent>> STP::getIntentsFromCurrentPlay(const World&
...
     updateSTPState(world);
-    return getIntentsFromCurrentPlay(world);
+    auto intents = getIntentsFromCurrentPlay(world);
+
+    auto all_tactics = stop_tactics;
+    all_tactics.push_back(goalie_tactic);
+    for (auto tactic : all_tactics)
+    {
+        auto iter = robot_tactic_assignment.find(tactic);
+        if (iter != robot_tactic_assignment.end())
```
Comment (tarakong01) [inline src/software/ai/hl/stp/stp.cpp:104]: Can you explain this if statement - why can't `iter` equal `robot_tactic_assignment.end()`?

File: src/software/ai/hl/stp/stp.cpp [hunk 2]
```diff
@@ -24,16 +24,22 @@ STP::STP(std::function<std::unique_ptr<Play>()> default_play_constructor,
...
       previous_override_play_name(""),
       override_play(false),
       previous_override_play(false),
-      current_game_state()
+      current_game_state(),
+      goalie_tactic(std::make_shared<GoalieTactic>(play_config->getGoalieTacticConfig())),
+      stop_tactics()
 {
+    for (unsigned int i = 0; i < MAX_ROBOT_IDS; i++)
+    {
```
Comment (tarakong01) [inline src/software/ai/hl/stp/stp.cpp:40]: What's the purpose of this loop?

File: src/software/ai/hl/stp/stp.cpp [hunk 3]
```diff
@@ -182,15 +206,21 @@ bool STP::overrideAIPlayIfApplicable()
...
-    std::map<std::shared_ptr<const Tactic>, Robot> robot_tactic_assignment;
+    robot_tactic_assignment.clear();
 
     std::optional<Robot> goalie_robot = world.friendlyTeam().goalie();
     std::vector<Robot> robots         = world.friendlyTeam().getAllRobots();
 
-    auto is_goalie_tactic = [](auto tactic) { return tactic->isGoalieTactic(); };
-    bool goalie_assigned  = false;
+    if (goalie_robot && automatically_assign_goalie)
+    {
```
Comment (tarakong01) [inline src/software/ai/hl/stp/stp.cpp:218]: In what game scenarios would `automatically_assign_goalie` be false?

File: src/software/ai/hl/stp/stp.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:104]: `iter != robot_tactic_assignment.end()` means that the tactic exists in the `robot_tactic_assignment` map

File: src/software/ai/hl/stp/stp.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:218]: example play and penalty kick

File: src/software/ai/hl/stp/stp.h [hunk 4]
```diff
@@ -136,14 +136,12 @@ class STP : public HL
...
      * {
-     *      {goalie_tactic},
      *      {crease_defender_1, crease_defender_2},
      *      {move_tactic},
      * }
      *
-     * The cost of assigning a goalie_tactic will be minimized across all robots first,
-     * followed by both the crease_defender tactics. The move_tactic will be assigned
-     * last.
+     * The cost of assigning borth the crease_defender tactics will be minimized across
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/stp.h]: borth -> both

File: src/software/ai/hl/stp/stp.cpp [hunk 5]
```diff
@@ -324,12 +335,5 @@ std::map<std::shared_ptr<const Tactic>, Robot> STP::assignRobotsToTactics(
         robots = remaining_robots;
     }
 
-    // store readable assignment map for PlayInfo
-    readable_robot_tactic_assignment.clear();
-    for (const auto& [tactic, robot] : robot_tactic_assignment)
-    {
-        readable_robot_tactic_assignment.emplace(robot.id(), objectTypeName(*tactic));
-    }
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/stp.cpp:332]: was this not being used anywhere?

**Review by akhilveeraghanta — COMMENTED** [review]
overall lgtm, pretty clean 

just 2 comments

File: src/software/ai/hl/stp/stp.cpp (hunk 5 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/stp.cpp:332]: `robot_tactic_assignment` is a member variable now, so we don't need to store `readable_robot_tactic_assignment` anymore

**Review by akhilveeraghanta — DISMISSED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Review by tarakong01 — APPROVED** [review]
lgtm!

**Pull request merged**